src/hierarchies/priors/abstract_prior_model.h

```diff
@@ -0,0 +1,96 @@
+#ifndef BAYESMIX_HIERARCHIES_PRIORS_ABSTRACT_PRIOR_MODEL_H_
+#define BAYESMIX_HIERARCHIES_PRIORS_ABSTRACT_PRIOR_MODEL_H_
+
+#include <google/protobuf/message.h>
+
+#include <random>
+#include <stan/math/prim.hpp>
+#include <vector>
+
+#include "algorithm_state.pb.h"
+#include "src/hierarchies/likelihoods/states/includes.h"
+#include "src/utils/rng.h"
+
+class AbstractPriorModel {
+ public:
+  //! Default destructor
+  virtual ~AbstractPriorModel() = default;
+
+  //! Returns an independent, data-less copy of this object. Implemented in
+  //! BasePriorModel
+  virtual std::shared_ptr<AbstractPriorModel> clone() const = 0;
+
+  //! Returns an independent, data-less deep copy of this object. Implemented
+  //! in BasePriorModel
+  virtual std::shared_ptr<AbstractPriorModel> deep_clone() const = 0;
+
+  //! Evaluates the log likelihood for the prior model, given the state of the
+  //! cluster
+  //! @param state_ A Protobuf message storing the current state of the cluster
+  //! @return The evaluation of the log likelihood
+  virtual double lpdf(const google::protobuf::Message &state_) = 0;
+
+  //! Evaluates the log likelihood for unconstrained parameter values.
+  //! By unconstrained parameters we mean that each entry of
+  //! the parameter vector can range over (-inf, inf).
+  //! Usually, some kind of transformation is required from the unconstrained
+  //! parameterization to the actual parameterization.
+  //! @param unconstrained_params vector collecting the unconstrained
+  //! parameters
+  //! @return The evaluation of the log likelihood of the prior model
+  virtual double lpdf_from_unconstrained(
+      Eigen::VectorXd unconstrained_params) const {
+    throw std::runtime_error("lpdf_from_unconstrained() not yet implemented");
+  }
+
+  //! Evaluates the log likelihood for unconstrained parameter values.
+  //! By unconstrained parameters we mean that each entry of
+  //! the parameter vector can range over (-inf, inf).
+  //! Usually, some kind of transformation is required from the unconstrained
+  //! parameterization to the actual parameterization. This version using
+  //! `stan::math::var` type is required for Stan automatic aifferentiation.
+  //! @param unconstrained_params vector collecting the unconstrained
+  //! parameters
+  //! @return The evaluation of the log likelihood of the prior model
+  virtual stan::math::var lpdf_from_unconstrained(
+      Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1> unconstrained_params)
+      const {
+    throw std::runtime_error(
+        "cluster_lpdf_from_unconstrained() not yet implemented");
+  }
+
+  //! Sampling from the prior model
+  //! @param use_post_hypers It is a `bool` which decides whether to use prior
+  //! or posterior parameters
```

Comment:
I would add that this is possible only for conjugate models

Reply:
Indeed, we now rely on the version receiving ``hier_hypers`` as input (plus a default behaviour that uses prior hyperparameters). Do we still need this method? I would update the docs and remove this version. Let me know what you think.

Reply:
I don't see the full picture, but if we don't need it, just remove it